tiproxy: update traffic replay docs (#20323)
pingcap/docs: tiproxy/tiproxy-traffic-replay.md

# TiProxy Traffic Replay

> Starting from TiProxy v1.3.0, you can use TiProxy to capture access traffic in a TiDB production cluster and replay it in a test cluster at a specified rate. This feature enables you to reproduce actual workloads from the production cluster in a test environment, verifying SQL statement execution results and performance. Starting from v1.4.0, TiProxy becomes generally available (GA).

Reviewer: ```suggestion
Starting from TiProxy v1.3.0, you can use TiProxy to capture access traffic in a TiDB production cluster and replay it in a test cluster at a specified rate. This feature enables you to reproduce actual workloads from the production cluster in a test environment, verifying SQL statement execution results and performance. Starting from v1.4.0, the TiProxy traffic replay feature becomes generally available (GA).
```

## Usage

> Before TiDB v9.0.0, only `tiproxyctl` is supported to connect to TiProxy for traffic capture and replay. Starting from TiDB v9.0.0, it is recommended to use SQL to capture and replay traffic.

Reviewer: ```suggestion
Before TiDB v9.0.0, only `tiproxyctl` is supported to connect to TiProxy for traffic capture and replay. Starting from TiDB v9.0.0, it is recommended to use SQL statements to capture and replay traffic.
```

<SimpleTab>
<div label="Use SQL">

1. Prepare the test environment:

    1. Create a test cluster. For more information, see [Deploy a TiDB Cluster Using TiUP](/production-deployment-using-tiup.md).
    2. Replicate data from the production cluster to the test cluster. For more information, see [Data Migration Overview](/migration-overview.md).
    3. Run the [`ANALYZE`](/sql-statements/sql-statement-analyze-table.md) statement in the test cluster to update statistics.

2. Use the [`TRAFFIC CAPTURE`](/sql-statements/sql-statement-traffic-capture.md) statement to capture traffic.

>     TiProxy supports capturing traffic to local and external storage. When capturing traffic to local, you need to manually copy the traffic file to the TiProxy cluster for replay after capturing the traffic, but when using external storage, there is no need to manually copy. TiProxy supports external storage including Amazon S3, Google Cloud Storage (GCS), Azure Blob Storage, or other file storage services that implement the S3 protocol. For more information about external storage, see [URI formats of external storage services](/external-storage-uri.md).

Reviewer: ```suggestion
    TiProxy supports capturing traffic to local and external storage. When capturing traffic to local, you need to manually copy the traffic file to the TiProxy cluster for replay after capturing the traffic, but when using external storage, there is no need to manually copy. TiProxy supports external storage including Amazon S3, Google Cloud Storage (GCS), Azure Blob Storage, or other S3-compatible file storage services. For more information about external storage, see [URI formats of external storage services](/external-storage-uri.md).
```

>     Capturing traffic requires the current user to have the `SUPER` or [`TRAFFIC_CAPTURE_ADMIN`](/privilege-management.md#dynamic-privileges) privilege.

Reviewer: ```suggestion
    To capture traffic, the current user must have the `SUPER` or [`TRAFFIC_CAPTURE_ADMIN`](/privilege-management.md#dynamic-privileges) privilege.
```

    > **Note:**
    >
    > - TiProxy captures traffic on all connections, including existing and newly created ones.
    > - The higher the CPU usage of TiProxy, the greater the impact of traffic capture on QPS. To reduce the impact on the production cluster, it is recommended to reserve at least 30% of CPU capacity, which results in an approximately 3% decrease in average QPS. For detailed performance data, see [Traffic capture test](/tiproxy/tiproxy-performance-test.md#traffic-capture-test).
    > - TiProxy does not automatically delete previous capture files when capturing traffic again. You need to manually delete them.

    For example, the following statement enables all TiProxy instances to capture traffic for one hour and save the traffic to the `/tmp/traffic` directory of each TiProxy instance:

    ```sql
    TRAFFIC CAPTURE TO "/tmp/traffic" DURATION="1h"
    ```

    Traffic files are automatically rotated and compressed. For more options, see [`TRAFFIC CAPTURE`](/sql-statements/sql-statement-traffic-capture.md).

3. If the traffic files are captured to the TiProxy local storage, copy the traffic file directory to the TiProxy instances in the test cluster.

> 4. Use [`TRAFFIC REPLAY`](/sql-statements/sql-statement-traffic-replay.md) statement to replay traffic.

Reviewer: ```suggestion
4. Use the [`TRAFFIC REPLAY`](/sql-statements/sql-statement-traffic-replay.md) statement to replay traffic.
```

    By default, SQL statements are executed at the same rate as in the production cluster, and each database connection corresponds to a connection in the production cluster to simulate the production load.

    For example, the following command connects to the TiProxy instance at `10.0.1.10:3080` using username `u1` and password `123456`, reads traffic files from the `/tmp/traffic` directory on the TiProxy instance, and replays the traffic:

    ```shell
    tiproxyctl traffic replay --host 10.0.1.10 --port 3080 --username="u1" --password="123456" --input="/tmp/traffic"
    ```

    Because all traffic runs under user `u1`, ensure `u1` can access all databases and tables. If no such user exists, create one. If the production cluster has a [resource group](/tidb-resource-control-ru-groups.md#manage-resource-groups), TiProxy automatically sets the resource group of each session to the same as when it was captured. Therefore, configure the [`SET RESOURCE GROUP`](/sql-statements/sql-statement-set-resource-group.md) [privilege](/sql-statements/sql-statement-set-resource-group.md#privilege) for `u1`.

>     If you replay all statements, before replaying again, you may need to restore the data to before the last replay to avoid errors caused by data duplication. You can also add the `--read-only=true` option to replay only read-only statements to avoid restoring data before each replay.

Reviewer 2: ```suggestion
    If you replay all statements, before replaying again, you might need to restore the data to before the last replay to avoid errors caused by data duplication. You can also add the `--read-only=true` option to replay only read-only statements to avoid restoring data before each replay.
```

To cancel the current capture or replay task, use the [`tiproxyctl traffic cancel`](/tiproxy/tiproxy-command-line-flags.md#traffic-cancel) command:
To cancel the current capture or replay job, use the [`tiproxyctl traffic cancel`](/tiproxy/tiproxy-command-line-flags.md#traffic-cancel) command:

```shell
tiproxyctl traffic cancel --host 10.0.1.10 --port 3080
```

For more information, see [`tiproxyctl traffic cancel`](/tiproxy/tiproxy-command-line-flags.md#traffic-cancel).

</div>
</SimpleTab>

## Limitations

- TiProxy only supports replaying traffic files captured by TiProxy and does not support other file formats. Therefore, make sure to capture traffic from the production cluster using TiProxy first.
- TiProxy traffic replay does not support filtering SQL types and DML and DDL statements are replayed. Therefore, you need to restore the cluster data to its pre-replay state before replaying again.
- TiProxy traffic replay does not support testing [Resource Control](/tidb-resource-control-ru-groups.md) and [privilege management](/privilege-management.md) because TiProxy uses the same username to replay traffic.

> - TiProxy does not support replaying [`LOAD DATA`](/sql-statements/sql-statement-load-data.md) statements.

Reviewer 2: 中文这一条没有删 https://github.com/pingcap/docs-cn/pull/19548/files#diff-230ff3de977941bfe7465bc1e0f533f93b27fa167729e233caf007f18df65f3aR301

Author: 哦，英文应该补上
